volatility3/framework/symbols/windows/extensions/__init__.py

```diff
@@ -994,6 +995,81 @@ def privileges(self):
                 vollog.log(constants.LOGLEVEL_VVVV, "Broken Token Privileges.")
 
 
+class KTIMER(objects.StructType):
+    """A class for Kernel Timers"""
+
+    VALID_TYPES = {
+        8: "TimerNotificationObject",
+        9: "TimerSynchronizationObject",
+    }
+
+    def get_signaled(self):
+        if self.Header.SignalState:
+            return "Yes"
+        return "-"
+
+    def get_raw_dpc(self):
+        """Returns the encoded DPC since it may not look like a pointer after encoding"""
+        symbol_table_name = self.get_symbol_table_name()
+        ulonglong_type = self._context.symbol_space.get_type(
+            symbol_table_name + constants.BANG + "unsigned long long"
+        )
+
+        return self._context.object(
+            object_type=ulonglong_type,
+            layer_name=self.vol.layer_name,
+            offset=self.Dpc.vol.offset,
+        )
+
+    def valid_type(self):
+        return self.Header.Type in self.VALID_TYPES
+
+    def get_due_time(self):
+        return "{0:#010x}:{1:#010x}".format(self.DueTime.HighPart, self.DueTime.LowPart)
+
+    def get_dpc(self):
+        """Return Dpc, and if Windows 7 or later, decode it"""
+        symbol_table_name = self.get_symbol_table_name()
+        kvo = self._context.layers[self.vol.native_layer_name].config[
+            "kernel_virtual_offset"
+        ]
+        ntkrnlmp = self._context.module(
+            symbol_table_name,
+            layer_name=self.vol.native_layer_name,
+            offset=kvo,
+            native_layer_name=self.vol.native_layer_name,
+        )
+
+        if ntkrnlmp.has_symbol("KiWaitNever") and ntkrnlmp.has_symbol("KiWaitAlways"):
+            wait_never = ntkrnlmp.object(
+                object_type="unsigned long long",
+                offset=ntkrnlmp.get_symbol("KiWaitNever").address,
+            )
+            wait_always = ntkrnlmp.object(
+                object_type="unsigned long long",
+                offset=ntkrnlmp.get_symbol("KiWaitAlways").address,
+            )
+
+            low_byte = (wait_never) & 0xFF
+            entry = utility.rol(self.get_raw_dpc() ^ wait_never, low_byte)
+            swap_xor = self.vol.offset | 0xFFFF000000000000
```

Comment:
Is this mask value specific to the way the code works, or is it a canonicalized address (in which case we have code for that) or is it based on the address space masking (ie, does it work if the processor decides to have more/fewer than normal bits for paging).

Reply:
@ikelos i believe it's just to get a canonicalized address.  the code from the other MR had a comment about doing it this way until something was merged... but I didn't see that that was merged. where is the code that could handle this rather than doing this `OR`

Reply:
think I found it, check the latest commit

Reply:
Sorry I didn't get back to you sooner, thanks for finding it!